tests/h/services/search_index/_queue_test.py

```diff
@@ -18,127 +20,132 @@
+    ):
+        annotation = factories.Annotation.create()
 
-class TestAddAnnotationsBetweenTimes:
-    @pytest.mark.parametrize("force", [True, False])
-    def test_it(self, annotation_ids, db_session, queue, force):
-        queue.add_annotations_between_times(
-            datetime_.datetime(2020, 9, 9),
-            datetime_.datetime(2020, 9, 11),
-            "test_tag",
-            force,
+        queue.add_where(
+            tag="test_tag",
+            priority=1,
+            where=[Annotation.id == annotation.id],
+            force=force,
         )
 
-        assert (
-            db_session.query(Job).all()
-            == Any.list.containing(
-                [
-                    Any.instance_of(Job).with_attrs(
-                        {
-                            "tag": "test_tag",
-                            "name": "sync_annotation",
-                            "scheduled_at": Any.instance_of(datetime_.datetime),
-                            "priority": 1000,
-                            "kwargs": {
-                                "annotation_id": str(
-                                    uuid.UUID(
-                                        URLSafeUUID.url_safe_to_hex(annotation_id)
-                                    )
-                                ),
-                                "force": force,
-                            },
-                        }
-                    )
-                    for annotation_id in annotation_ids
-                ]
-            ).only()
+        assert db_session.query(Job).one() == Any.instance_of(Job).with_attrs(
+            {
+                "kwargs": Any.dict.containing(
+                    {
+                        "force": expected_force,
+                    }
+                )
+            }
         )
 
-    @pytest.fixture
-    def annotations(self, factories):
-        return factories.Annotation.create_batch(
-            size=2, updated=datetime_.datetime(year=2020, month=9, day=10)
+    def test_add(self, queue, add_where):
+        queue.add(
+            sentinel.annotation_id,
+            sentinel.tag,
+            schedule_in=sentinel.schedule_in,
+            force=sentinel.force,
         )
 
-    @pytest.fixture(autouse=True)
-    def non_matching_annotations(self, factories):
-        """Annotations from outside the date range that we're reindexing."""
-        factories.Annotation.create(
-            updated=datetime_.datetime(year=2020, month=9, day=8)
+        add_where.assert_called_once_with(
+            sentinel.tag,
+            Queue.Priority.SINGLE_ITEM,
+            [Any.instance_of(BinaryExpression)],
+            sentinel.force,
+            sentinel.schedule_in,
         )
-        factories.Annotation.create(
-            updated=datetime_.datetime(year=2020, month=9, day=12)
+
+        where = add_where.call_args[0][2]
+        assert where[0].compare(Annotation.id == sentinel.annotation_id)
```

Comment:
SQLAlchemy `BinaryExpression` objects are a bit of a pain to match, as I think they are intended to return True when they match the actual situation they express, rather than when they have the same arguments as another one.

This means you can make two identical `BinaryExpressions` and they aren't naturally equal. They do have a `compare()` method that does that, but this means we can't use any fancy matcher stuff with them without writing a custom matcher or something.

Reply:
I think we might be better off not mocking `add_where()` in these tests, and just checking for the jobs created in the DB.

Each `add_*()` method can have only the tests that it needs and indirectly test only the aspects of `add_where()` that it cares about, then `add_where()`'s own tests can thoroughly test all aspects of `add_where()`. This means some duplication in terms of tests re-testing the same aspects of `add_where()` but I think it'd be overall better.
I'm generally not a fan of the unit tests for a class mocking out parts of that same class, and in this case it's also getting us dangerously close to mocking SQLAlchemy which is why we're having trouble. It means the tests know less about the internal implementation details of `Queue` (they no longer know that all the `add_*()` methods are implemented on top of `add_where()`), means the tests are more black-box and end-to-end, and actually makes the tests nicer, particularly because these assertions get much better as they can be done in a one-er with h-matchers.

Example:

```diff
diff --git a/tests/h/services/search_index/_queue_test.py b/tests/h/services/search_index/_queue_test.py
index effef9cd7..d353901e6 100644
--- a/tests/h/services/search_index/_queue_test.py
+++ b/tests/h/services/search_index/_queue_test.py
@@ -85,25 +85,27 @@ class TestAddMethods:
             }
         )
 
-    def test_add(self, queue, add_where):
+    def test_add(self, db_session, factories, queue):
+        annotation = factories.Annotation()
+
         queue.add(
-            sentinel.annotation_id,
-            sentinel.tag,
-            schedule_in=sentinel.schedule_in,
-            force=sentinel.force,
+            annotation.id,
+            "test_tag",
+            schedule_in=30,
+            force=True,
         )
 
-        add_where.assert_called_once_with(
-            sentinel.tag,
-            Queue.Priority.SINGLE_ITEM,
-            [Any.instance_of(BinaryExpression)],
-            sentinel.force,
-            sentinel.schedule_in,
+        assert db_session.query(Job).one() == Any.instance_of(Job).with_attrs(
+            {
+                "tag": "test_tag",
+                "priority": 1,
+                "kwargs": {
+                    "annotation_id": self.mapped_id(annotation),
+                    "force": True,
+                },
+            }
         )
 
-        where = add_where.call_args[0][2]
-        assert where[0].compare(Annotation.id == sentinel.annotation_id)
-
     def test_add_annotations_between_times(self, queue, add_where):
         queue.add_annotations_between_times(
             sentinel.start_time, sentinel.end_time, sentinel.tag, force=sentinel.force
```